designs/validations.md

```diff
@@ -0,0 +1,288 @@
+
+The rest of validations and defaults (static data for a group of API structs and the context aware ones) **can** live anywhere, as long the dependency directions make sense, with a few constrains:
+* Context aware validation **cannot** live in `pkg/api/v1**`. This is to keep this package dependencies at a minimum.
+* **Avoid** packages like `validations`. Don't group validations for what they are, group them by what they validate (their context). For example, for docker validations, don't use `validations.ValidateDockerDesktopVersion`, think about `docker.ValidateDesktopVersion` in `pkg/docker`.
+* **Avoid** the `cmd` package. These tend to end up being untested and lack structure.
+
+Webhooks that only require static data validations/defaults **should** live in the `pkg/api/v1**` (where all present webhooks live today). If they have any other dependency the **cannot** live in the `api` package. (*Note*: we could possibly define a `webhooks` folder elsewhere if we think that could avoid problems in the future, leaving this up for discussion)
+
+### Scope (*what*)
+#### Defaults
+* Defaults should not **modify** data provided by the user, they can only **add** information. Any transformation needed should be moved to runtime operations and never get reflected back on the API objects.
+* They should have not secondary effects on anything else except the provided API object/s, including the internal state of the entity running the validation.
+* API objects should only be modified by the defaulting logic. This guarantees they are invariants during the CLI execution, allowing for workflow reentry/re-execution (think command fails and on the next execution we continue from where it stopped).
+
+#### Validations
+* They should have not secondary effects on **anything**, including the internal state of the entity running the validation.
+* API struct `Validate` methods should return a [`field.ErrorList`](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/validation/field#ErrorList). This facilitates using it from webhooks and ensures the API user facing error is as complete as possible.
+
+### Calling (*when*)
+#### Defaults
+ * CLI
+	* At the beginning of the command flow, in most cases before any other business logic.
+	* Always before validations.
+	* This doc doesn't prescribe which entity should trigger the defaults (command in `cmd` package, workflow, etc.). But they do need to be run before the API objects get injected in any other entity as a dependency (think *service* objects). Given the approach the provider's refactor project has taken to inject the `cluster.Spec` and any other non dynamic data (this is, known at the beginning of the command execution) as state in the service objects, `cmd`  seems like the obvious place to do this since this is where service objects are constructed.
+* API
```

Comment:
Is this also before respective validations in webhook and controllers or not the same as cli?

Reply:
Sorry, I don't follow. The API bullet point only refers to "webhooks + controllers"